fireplace/cards/utils.py

```diff
@@ -39,52 +41,40 @@
 
 
 # take a list of weights and a list of card pools and produce a random weighted sample without replacement
-def weighted_card_choice(source, weights, card_sets, count):
+# len(weights) == len(card_sets) (one weight per card set)
+def weighted_card_choice(source, weights: List[int], card_sets: List[str], count: int):
 
 	# special case for single weight (generic random pick)
 	if len(card_sets)==1:
 		return [source.controller.card(card, source=source) for card in random.sample(card_sets[0], count)]
 
-	# get all the cards matching each set of filters and create the weight line
-	# (user does not have to normalize weighting from 0-1)
-	cardlist = []
-	weightedsegmentstarts = [0] * len(card_sets)
-	weightlinelength = 0
+	# special case for all weights equal
+	if weights.count(weights[0]) == len(weights):
+		card_set = [j for i in card_sets for j in i]
+		return [source.controller.card(card, source=source) for card in random.sample(card_set, count)]
 
-	for i, c in enumerate(card_sets):
-		weightedsegmentstarts[i] = len(cardlist)
-		cardlist += c
-		# higher weighted sets (segments) occupy more space per item on the weight line
-		weightlinelength += len(c) * weights[i]
+	chosen_cards = []
 
-	ret = []
+	# sum all the weights
+	cum_weights = []
+	totalweight = 0
+	for i, w in enumerate(weights):
+		totalweight += w * len(card_sets[i])
+		cum_weights.append(totalweight)
 
 	# for each card
-	for c in range(count):
-
-		# pick a point along the weight line
-		r = random.random() * weightlinelength
-		pos = 0
-
-		# figure out which segment it's in
-		for i, w in enumerate(weights):
-			segmentarraylength = ((len(cardlist) if i == len(weights)-1 else weightedsegmentstarts[i+1]) - weightedsegmentstarts[i])
-			segmentlinelength = segmentarraylength * w
-			# it's in this segment, find exact card
-			if (r < pos + segmentlinelength):
-				# find percentage along this segment we're at
-				spc = (r - pos) / segmentlinelength
-				# map percentage to array index, add card to selection, remove from weight line
-				ret.append(cardlist.pop(weightedsegmentstarts[i] + int(spc * segmentarraylength)))
-				# update weights
-				weightlinelength -= w
-				weightedsegmentstarts[i+1:] = [x-1 for x in weightedsegmentstarts[i+1:]]
-
-				break
-
-			pos += segmentlinelength
-
-	return [source.controller.card(card, source=source) for card in ret]
+	for i in range(count):
+		# choose a set according to weighting
+		chosen_set = bisect(cum_weights, random.random() * totalweight)
```

Comment:
Yes, I'm just always leery of mixed-mode math because every language seems to have its own quirks.

I'm fine with making the weights floats, but then we should change the type signature